Userland/Utilities/pls.cpp

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2021 Jesse Buhagiar <jooster669@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <AK/Types.h>
+#include <AK/Vector.h>
+#include <LibCore/Account.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <LibCore/GetPassword.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+static constexpr mode_t EXPECTED_PERMS = (S_IFREG | S_IRUSR);
+
+// Function Definitions
+extern "C" int main(int arch, char** argv);
+int unveil_paths(const char*);
+
+// Unveil paths, given the current user's path and the command they want to execute
+int unveil_paths(const char* command)
+{
+    int num_unveils = 0;
+
+    // Attempt to unveil the command first...
+    auto* command_path = realpath(command, nullptr);
+    dbgln(command_path);
+    if (command_path) {
+        if (unveil(command_path, "x") == 0)
+            num_unveils++;
+
+        free(command_path);
+        return num_unveils;
+    }
+
+    // Okay, so we couldn't find the actual file specified by the user, let's
+    // instead search PATH for it...
+    auto command_path_system = Core::find_executable_in_path(command);
```

Comment:
Sorry I don't quite understand what you mean here. Will `Core::find_executable_in_path()` not work before I unveil directories in the path? `which` seems to use this without any calls to unveil? Unveil is only ever gets called once here. 

The first check in this function is to check for a local command, i.e `pls ./sdlquake2` that can be resolved via `realpath`. If that fails, we then use `find_executable_in_path` to get the full path of the command typed in, if any, as something in the path has probably been specified, i.e `pls whoami`, `pls ls`. Perhaps I'm misunderstanding the correct order of operations that should be occurring here. 

Also the return value/use of `num_unveils` is misleading. I've changed this to `bool` to signify that it either unveils the command or it doesn't, meaning it's probably non-existent.

Reply:
After you have called `unveil()` at least once, your view of file system is now constrained to whatever paths you have unveiled. In this program, you first `unveil` a handful of paths in `main()`, and then call `unveil_paths()`.

This means that `Core::find_executable_in_path()` cannot see into `/bin`, for example. Since the process has a file system veil, the only syscall that can see `/bin` is a call to `unveil()` to unveil `/bin`. If you do that, `/bin` is now visible to all other syscalls as well.

Reply:
So it would be best to `unveil` the path to the program first, unveil `/etc/` to unveil the files required (`plsusers`, `passwd` etc), then lock the veil?